### PR TITLE
Adding OPTForSeqClassification class

### LINKED ISSUE #17525: Is the addition of the 'OPTforSequenceClassification' class scheduled?
### Feature request

Is the addition of the 'OPTforSequenceClassification' class scheduled? 
Is someone handling it? 
When adding these functions, I wonder if it is possible to PR one by one, or if I have to PR all classes supported by other models.

### Motivation

Added function of OPT class, which is being actively discussed recently

### Your contribution

I personally use the forSequenceClassification class because I need it for my experiments, but I would like to inquire if I can contribute to the addition of the function.

### PR DESCRIPTION
**Pull request by oneraghavan**
# What does this PR do?
It add the class for OPTForSequenceClassification based on OPT model 

Fixes #17525 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by oneraghavan** [comment]
@NielsRogge Can you help me on how to fix the tests?

File: src/transformers/models/opt/modeling_opt.py [hunk 1]
```diff
@@ -38,12 +38,11 @@
 
 _CHECKPOINT_FOR_DOC = "facebook/opt-350m"
 _CONFIG_FOR_DOC = "OPTConfig"
-_TOKENIZER_FOR_DOC = "GPT2Tokenizer"
+_TOKENIZER_FOR_DOC = "OPTTokenizer"
```
Comment (LysandreJik) [inline src/transformers/models/opt/modeling_opt.py:41]: That was correct, the tokenizer that should be used is `GPT2Tokenizer`.

File: src/transformers/models/opt/modeling_opt.py [hunk 2]
```diff
@@ -38,12 +38,11 @@
...
 
 # Base model docstring
 _EXPECTED_OUTPUT_SHAPE = [1, 8, 1024]
 
-
```
Comment (LysandreJik) [inline src/transformers/models/opt/modeling_opt.py:46]: Please revert this change

File: src/transformers/models/opt/modeling_opt.py [hunk 3]
```diff
@@ -762,7 +761,6 @@ def forward(
         output_hidden_states: Optional[bool] = None,
         return_dict: Optional[bool] = None,
     ) -> Union[Tuple, BaseModelOutputWithPast]:
-
```
Comment (LysandreJik) [inline src/transformers/models/opt/modeling_opt.py:765]: same here

File: src/transformers/models/opt/modeling_opt.py [hunk 4]
```diff
@@ -802,7 +800,7 @@ def __init__(self, config):
         self.model = OPTModel(config)
 
         # the lm_head weight is automatically tied to the embed tokens weight
-        self.lm_head = nn.Linear(config.word_embed_proj_dim, config.vocab_size, bias=False)
+        self.lm_head = nn.Linear(config.embed_dim, config.vocab_size, bias=False)
```
Comment (LysandreJik) [inline src/transformers/models/opt/modeling_opt.py:803]: This seems like a fix that isn't linked to the sequence classification head contribution; if that's so, could you split it in a different PR where we can discuss it? Thanks.

File: src/transformers/models/opt/modeling_opt.py [hunk 5]
```diff
@@ -976,3 +974,134 @@ def _reorder_cache(past, beam_idx):
...
+
+    def __init__(self, config: OPTConfig):
+        super().__init__(config)
+        self.num_labels = config.num_labels
+        self.transformer = OPTModel(config)
+        self.score = nn.Linear(config.word_embed_proj_dim, self.num_labels, bias=False)
+
+        # Model parallel
+        self.model_parallel = False
+        self.device_map = None
```
Comment (LysandreJik) [inline src/transformers/models/opt/modeling_opt.py:1005]: this can be removed
```suggestion
```

**Review by LysandreJik — COMMENTED** [review]
Thanks for your PR! Added a few comments and pinged @ArthurZucker as he has more direct experience with the model to vouch for the sequence classification head more specifically

**Comment by oneraghavan** [comment]
@ArthurZucker  Please point on how to fix these errors

File: docs/source/en/model_doc/opt.mdx [hunk 6]
```diff
@@ -44,6 +44,11 @@ The original code can be found [here](https://github.com/facebookresearch/metase
 [[autodoc]] OPTForCausalLM
     - forward
 
+## OPTForSequenceClassification
+
+[[autodoc]] OPTForSequenceClassification
+    - forward
+
```
Comment (ArthurZucker) [inline docs/source/en/model_doc/opt.mdx:51]: Would probably just place this after ## TFOPTForCausalLM

File: src/transformers/models/opt/modeling_opt.py [hunk 7]
```diff
@@ -976,3 +976,130 @@ def _reorder_cache(past, beam_idx):
...
+        self.post_init()
+
+    @add_start_docstrings_to_model_forward(OPT_INPUTS_DOCSTRING)
+    @add_code_sample_docstrings(
+        processor_class=_TOKENIZER_FOR_DOC,
+        output_type=SequenceClassifierOutputWithPast,
+        config_class=_CONFIG_FOR_DOC,
+        expected_output="'LABEL_0'",
+        expected_loss=5.28,
+    )
```
Comment (ArthurZucker) [inline src/transformers/models/opt/modeling_opt.py:1015]:
```suggestion
    )
    def get_input_embeddings(self):
        return self.transformer.decoder.embed_tokens

    def set_input_embeddings(self, value):
        self.transformer.decoder.embed_tokens = value
```
These two functions are mostly used for testing and could probably solve the 2 failing tests

File: src/transformers/models/opt/modeling_opt.py [hunk 8]
```diff
@@ -976,3 +976,130 @@ def _reorder_cache(past, beam_idx):
...
+    def forward(
+        self,
+        input_ids: Optional[torch.LongTensor] = None,
+        past_key_values: Optional[Tuple[Tuple[torch.Tensor]]] = None,
```
Comment (ArthurZucker) [inline src/transformers/models/opt/modeling_opt.py:1019]: Small nit, but we usually put the past key value argument after the head mask ( see OPTForCausalLM)

File: src/transformers/models/opt/modeling_opt.py [hunk 9]
```diff
@@ -976,3 +976,130 @@ def _reorder_cache(past, beam_idx):
...
+        logits = self.score(hidden_states)
+
+        if input_ids is not None:
+            batch_size, sequence_length = input_ids.shape[:2]
+        else:
+            batch_size, sequence_length = inputs_embeds.shape[:2]
+
+        assert (
+            self.config.pad_token_id is not None or batch_size == 1
+        ), "Cannot handle batch sizes > 1 if no padding token is defined."
```
Comment (ArthurZucker) [inline src/transformers/models/opt/modeling_opt.py:1058]: We don't put assert in the code, this should be removed 🤗 Or replace with a raise value error

File: tests/models/opt/test_modeling_opt.py [hunk 10]
```diff
@@ -242,6 +245,18 @@ def test_generate_fp16(self):
         model.generate(input_ids, attention_mask=attention_mask)
         model.generate(num_beams=4, do_sample=True, early_stopping=False, num_return_sequences=3)
 
+    def test_gpt2_sequence_classification_model(self):
```
Comment (ArthurZucker) [inline tests/models/opt/test_modeling_opt.py:248]: should be renamed to `test_opt_sequence_classification_model

File: src/transformers/models/opt/modeling_opt.py [hunk 11]
```diff
@@ -729,7 +729,7 @@ class OPTModel(OPTPreTrainedModel):
     def __init__(self, config: OPTConfig):
         super().__init__(config)
         self.decoder = OPTDecoder(config)
-
+        self.embed_dim = config.hidden_size
```
Comment (ArthurZucker) [inline src/transformers/models/opt/modeling_opt.py:732]: Is this really needed? (I am wondering as it not in the main version )

**Review by ArthurZucker — CHANGES_REQUESTED** [review]
LGTM, thanks for your contribution. 

Would love to have your review on that @younesbelkada :)

**Review by younesbelkada — APPROVED** [review]
Thanks a lot for your nice contribution 🚀 
Added a small comment, would be nice to add another test for `multi_label_classification`mode :)

**Comment by oneraghavan** [comment]
@younesbelkada Some tests are failing due to same model(OPTForSeqClassification) available in torch not available for tensorflow. Should I add it part of this PR ? 

Can we close this PR and I will raise another PR with TFOPTForSeqClassification ?

**Comment by ArthurZucker** [comment]
Hey, I think the history is a bit messed up but it is alright (issues with merging I guess). Also, we should not have to change the `OPTPretrainedModel` base prefix. This can be a backward compatibility issue and should definitely be avoided. If you could just revert on that change it would be great. I think that it will solve the failing test and we will be able to merge 👍🏻

File: src/transformers/models/opt/modeling_opt.py [hunk 12]
```diff
@@ -383,8 +383,9 @@ def forward(
     OPT_START_DOCSTRING,
 )
 class OPTPreTrainedModel(PreTrainedModel):
+
     config_class = OPTConfig
-    base_model_prefix = "model"
+    base_model_prefix = "transformer"
```
Comment (ArthurZucker) [inline src/transformers/models/opt/modeling_opt.py:388]:
```suggestion
    base_model_prefix = "model"
```

Comment (oneraghavan) [reply src/transformers/models/opt/modeling_opt.py:388]: I changed the base_model_prefix to transformers as OPTModelTest::test_load_with_mismatched_shapes was failing with base_model_prefix = "model" .

**Comment by oneraghavan** [comment]
> Hey @oneraghavan ! Thanks for your comment, I am not sure if not having `TFOPTForSeqClassification` explains why those tests are failing since the test should not return nothing [here](https://github.com/oneraghavan/transformers/blob/d672d9c54a5a329220a8dabb6b6e3f961fbdca5b/tests/test_modeling_common.py#L1767). The error says `AttributeError: decoder.embed_tokens.weight not found in TF 2.0 model` so it might be possible that the modifications you made on `OPTModel` and `OPTPretrainedModel` classes broke those tests. Also the git history seems to be broken, could you please rebase to main or merge with force-push to clean the git histoiry (aka the number of modified files has increased) 💪 Thanks again for your help here! And let us know if anything is unclear or if you need any help

I Tried to do pull rebase from my main to huggingface repo. How do you want me to leave this PR? Just leave my commits on top ?

**Comment by oneraghavan** [comment]
@ArthurZucker All tests fixed, we can close this PR.

**Review by LysandreJik — APPROVED** [review]
Looks good! Only left a comment regarding one addition.

**Comment by oneraghavan** [comment]
> Looks good! Only left a comment regarding one addition.

Done

**Review by ArthurZucker — APPROVED** [review]
LGTM let's merge! Good job @oneraghavan 👍🏻 Thanks for the contribution

**Comment by ArthurZucker** [comment]
Hey @oneraghavan do you have any checkpoints we could use for the documentation? It seems that you added some expected loss value and expected outputs, wanted to know if we can maybe use the model checkpoints for documentation (otherwise it is totally ok! I will use a dummy model 😄 )

**Comment by oneraghavan** [comment]
@ArthurZucker  I do not have any checkpoints, I used a dummy model.
